File: e2e/factories/companies.ts
import { faker } from "@faker-js/faker";
import { db } from "@test/db";
import { companyAdministratorsFactory } from "@test/factories/companyAdministrators";
import { companyStripeAccountsFactory } from "@test/factories/companyStripeAccounts";
import { usersFactory } from "@test/factories/users";
import { companies } from "@/db/schema";
import { assert } from "@/utils/assert";

type CreateOptions = {
  withoutBankAccount: boolean;
};

export const companiesFactory = {
  create: async (
    overrides: Partial<typeof companies.$inferInsert> = {},
    options: CreateOptions = { withoutBankAccount: false },
  ) => {
    const [insertedCompany] = await db
      .insert(companies)
      .values({
        name: faker.company.name(),
        email: faker.internet.email(),
        registrationNumber: faker.string.numeric(9),
        registrationState: "DE",
        streetAddress: faker.location.streetAddress(),
        city: faker.location.city(),
        state: faker.location.state({ abbreviated: true }),
        zipCode: faker.location.zipCode(),
        countryCode: "US",
        stripeCustomerId: "cus_M2QFeoOFttyzTx",
        brandColor: faker.color.rgb({ prefix: "#" }),
        website: "https://www.example.com",
        requiredInvoiceApprovalCount: 2,
        fullyDilutedShares: 1000000n,
        valuationInDollars: 2000000n,
        sharePriceInUsd: "100",
        fmvPerShareInUsd: "40",
        ...overrides,
      })
      .returning();
    assert(insertedCompany != null);

    if (!options.withoutBankAccount && !overrides.stripeCustomerId) {
      await companyStripeAccountsFactory.create({ companyId: insertedCompany.id });
    }

    return { company: insertedCompany };
  },

  createPreOnboarding: async (
    overrides: Partial<typeof companies.$inferInsert> = {},
    options: CreateOptions = { withoutBankAccount: true },
  ) =>
    companiesFactory.create(
      {
        name: null,
        streetAddress: null,
        city: null,
        state: null,
        zipCode: null,
        stripeCustomerId: null,
        ...overrides,
      },
      options,
    ),

  createCompletedOnboarding: async (
    overrides: Partial<typeof companies.$inferInsert> = {},
    options: CreateOptions = { withoutBankAccount: false },
  ) => {
    const company = (await companiesFactory.create(overrides, options)).company;
    const adminUser = (await usersFactory.create()).user;
    await companyAdministratorsFactory.create({
      companyId: company.id,
      userId: adminUser.id,
    });
    return { company, adminUser };
  },
};

File: backend/spec/factories/companies.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :company do
    name { Faker::Company.name }
    email { Faker::Internet.unique.email }
    registration_number { Faker::Company.duns_number }
    registration_state { "DE" }
    street_address { Faker::Address.street_address }
    city { Faker::Address.city }
    state { Faker::Address.state_abbr }
    zip_code { Faker::Address.zip_code }
    country_code { "US" }
    stripe_customer_id { "cus_M2QFeoOFttyzTx" }
    brand_color { Faker::Color.hex_color }
    website { "https://www.example.com" }
    required_invoice_approval_count { 2 }
    fully_diluted_shares { 1_000_000 }
    valuation_in_dollars { 2_000_000 }
    share_price_in_usd { 100 }
    fmv_per_share_in_usd { 40 }

    transient do
      without_bank_account { false }
    end

    after :build do |company, evaluator|
      company.bank_account = build(:company_stripe_account, company:) unless company.bank_account.present? || evaluator.without_bank_account
    end

    trait :without_bank_account do
      without_bank_account { true }
      stripe_customer_id { nil }
    end

    trait :pre_onboarding do
      without_bank_account
      name { nil }
      street_address { nil }
      city { nil }
      state { nil }
      zip_code { nil }
    end

    trait :completed_onboarding do
      after :create do |company|
        create(:company_administrator, company:)
      end
    end

    trait :with_logo do
      after :create do |company|
        company.logo.attach(io: File.open(Rails.root.join("spec", "fixtures", "files", "company-logo.png")), filename: "company-logo.png")
      end
    end
  end
end
